sound/soc/sof/intel/hda.c

```diff
@@ -1107,7 +1107,16 @@ static int hda_sdw_machine_select(struct snd_sof_dev *sdev)
 	if (link_mask && !pdata->machine) {
 		for (mach = pdata->desc->alt_machines;
 		     mach && mach->link_mask; mach++) {
-			if (mach->link_mask != link_mask)
+			/*
+			 * On some platforms such as Up Extreme all links
+			 * are enabled but only one link can be used by
+			 * external codec. Instead of exact match of two masks,
+			 * first check whether link_mask of mach is subset of
+			 * link_mask supported by hw and then go on searching
+			 * link_adr
+			 */
+			if (!mach->link_mask || !link_mask ||
```

Comment:
the test for `!mach->link_mask` is redundant - you wouldn't enter the loop in such a case. So the test for `!link_mask` is redundant too: if `!link_mask` then `~link_mask == ~0` and since `mach->link_mask != 0` then `~link_mask & mach->link_mask != 0`. So, you only need to test `if (~link_mask & mach->link_mask)`

Reply:
@lyakh thanks, I totally agree with you.